conftest.py

```diff
@@ -1,14 +1,330 @@
+    template.
+
+    If sampling is an int, the number of parameter tuples returned will
+    be (sampling) for each template.
+
+    If sampling is "exhaustive", we will return all possible parameter
+    tuples for each template.
+
+    If sampling is "bucket", we will yield parameter tuples as follows:
+        For each template parameter name:
+            Randomly sample one template parameter value for all other
+                template parameter names
+            Yield the pytest param for the set of template parameter
+    """
+    if seed is not None:
+        random.seed(seed)
+    parameter_tuples = []
+    current_module_dir = pathlib.Path(__file__).parent.absolute()
+    template_files = (
+        os.path.join(root, f)
+        for root, _, files in os.walk(current_module_dir, followlinks=True)
+        for f in files
+        if f.endswith(".template.mlir")
     )
+    for template_file in template_files:
+        # Parse the template file
+        with open(template_file, "r") as f:
+            json_sting, mlir_template = f.read().split("### START TEST ###")
+        test_spec: dict = json.loads(json_sting)
+        mlir_template = jinja2.Template(mlir_template, undefined=jinja2.StrictUndefined)
+        if "parameters" not in test_spec:
+            raise ValueError(
+                f"{template_file} does not contain a valid test specification as "
+                'it does not specify a value for the key "parameters".'
+            )
+        elif "run" not in test_spec or not isinstance(test_spec["run"], str):
+            raise ValueError(
+                f"{template_file} does not contain a valid test specification as "
+                'it does not specify a valid value for the key "run".'
+            )
+        prefilter_name = test_spec.get("prefilter")
+        parameter_dict_filter = PREFILTER_FUNCTIONS.get(prefilter_name)
+        if parameter_dict_filter is None:
+            raise NameError(f"Unknown prefilter function named {repr(prefilter_name)}.")
+
+        # Grab test running command
+        test_execution_command_template = test_spec["run"]
+
+        # Grab parameter choices
+        parameter_choices: Dict[str, List[str]] = dict()
+        for parameter_name, parameter_value_choices in test_spec["parameters"].items():
+            if (
+                isinstance(parameter_value_choices, str)
+                and parameter_value_choices in NAMED_PARAMETER_VALUE_CHOICES
+            ):
+                parameter_choices[parameter_name] = NAMED_PARAMETER_VALUE_CHOICES[
+                    parameter_value_choices
+                ]
+            elif isinstance(parameter_value_choices, list):
+                parameter_choices[parameter_name] = parameter_value_choices
+            else:
+                raise ValueError(
+                    f"{repr(parameter_value_choices)} does not specify a valid set of parameter values."
+                )
+
+        # Handle each sampling case separately
```

Comment:
Which sampling method matches the existing behavior of one value per parameter with a predictable choice?
`sampling=1` seems the closest, but the choice of the value is random.

Reply:
I don't think I have one for what you're describing. I can add one if desired. 

I'll summarize what I have so far.

Bucket:
```
for each template
    for each parameter
        for each possible value this parameter can take:
            randomly pick exactly one value for all the other parameters
            Make a test
```

Exhaustive:
```
for each template
    grab all possible tests for this template
    add all the possible tests to the final tests
```

If `sampling` is an `int`:
```
for each template:
    grab all possible tests for this template
    randomly "sampling" number of tests from all possible tests (add these to the final tests)
```

If `sampling` is a `float` in the range [0,1]:
```
for each template:
    grab all possible tests for this template
    randomly "sampling" percent of all possible tests (add these to the final tests)
```

I think we can have one for what you're describing and call it `default` that'll work like this:
```
for each template:
    for each parameter
        pick the first choice (deterministic since the parameter choices are lists).
    add exactly one test for this template
```

Reply:
I like `default` for the name. Can we also make that the default method of running the tests if not specified?

Reply:
I've implemented the `default` option and made it the default method of running tests. I've broken the tests somewhere in there, so I'll be fixing that.